### PR TITLE
[WIP] Set default storage class on PVC sync if default StorageClass is present

### LINKED ISSUE #99629: Default storage class needs to be present before PVCs are created
#### What happened:

When the PVC is created before the default storage class, PVC will end up in a state with no storage class and will never be bounded to a PV.

#### What you expected to happen:

I expect that in eventually consistent model of Kubernetes, non-bound PVCs with no storage class will get assigned the default class once one is available and get bounded to a PV.

#### How to reproduce it (as minimally and precisely as possible):

Create a PVC with no storage class, and then create a StorageClass marked as default.

#### Anything else we need to know?:

The setter of the default storage class on PVCs is currently implemented only as an admission webhook, creating the requirements for the default StorageClass to exist before any PVCs are created. I think similar logic should be executed in PV controller to add the default storage class if PVC is not bound and doesn't have a storage class.

#### Environment:
- Kubernetes version (use `kubectl version`): 1.19.4

### PR DESCRIPTION
**Pull request by LittleChimera**
In case where there's no default StorageClass at the time of creation of a PVC, the PVC will stay without storage class.
This logic was thus far only executed in AdmissionWebhook which would add a storageClassName to PVC with no defined storageClass.
Therefore, PVC will stay without storageClassName, if the PVC manifest is not reapplied to the cluster.

#### Which issue(s) this PR fixes:

Fixes #99629

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 1]
```diff
@@ -359,8 +360,19 @@ func (ctrl *PersistentVolumeController) syncUnboundClaim(claim *v1.PersistentVol
...
+			case storagehelpers.GetPersistentVolumeClaimClass(claim) == "":
+				if storageClasses, err := ctrl.kubeClient.StorageV1().StorageClasses().List(context.TODO(), metav1.ListOptions{}); err != nil {
+					for _, sc := range storageClasses.Items {
+						if sutil.IsDefaultAnnotation(sc.ObjectMeta) {
+							claim.Spec.StorageClassName = &sc.Name
+							ctrl.updateClaim(claim)
+							return nil
+						}
+					}
+				}
```
Comment (jsafrane) [inline pkg/controller/volume/persistentvolume/pv_controller.go:372]: Please extract all this getting default SC and modifying PV into a separate function and just call it. I like the giant `if-then-else` in `syncUnboundClaim` as clean as possible.

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 2]
```diff
@@ -359,8 +360,19 @@ func (ctrl *PersistentVolumeController) syncUnboundClaim(claim *v1.PersistentVol
 					return err
 				}
 				return nil
+			case storagehelpers.GetPersistentVolumeClaimClass(claim) == "":
+				if storageClasses, err := ctrl.kubeClient.StorageV1().StorageClasses().List(context.TODO(), metav1.ListOptions{}); err != nil {
```
Comment (jsafrane) [inline pkg/controller/volume/persistentvolume/pv_controller.go:364]: Setting the default storage class should be done at line 337, "// User did not care which PV they get" - if the PVC has no storage class, file it there and return. Updated PVC will be enqueued shortly and it will have a SC.

**Review by jsafrane — COMMENTED** [review]
Please add some unit tests too!

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 2 above)
Comment (jsafrane) [reply pkg/controller/volume/persistentvolume/pv_controller.go:364]: Hold on, maybe leave the code here - if there is a PV without a storage class, we try to bind to it once. If there is no such PV, we file the default SC into the PVC and try again, hoping for dynamic provisioning. This will keep at least a bit of the existing behavior there.

@msau42 what do you think? It's a bit confusing.

**Comment by thockin** [comment]
Is this really something we want to do?  If I recall correctly, having no
class is semantically meaningful, isn't it?

**Comment by LittleChimera** [comment]
Thank you for the feedback! ❤️ 

> Is this really something we want to do? If I recall correctly, having no class is semantically meaningful, isn't it?
> […](#)

> Please add some unit tests too!

I was actually trying to modify these test cases which specify that nothing should be done when storageClass is set to an empty value (with this PR, I would expect that in these two cases, the default storageClass is overridden on this PVC instead, and then a PV provisioned and bound):
https://github.com/jsafrane/kubernetes/blob/bb5d562f37ab029d2b562e23bd23e906ee099a45/pkg/controller/volume/persistentvolume/provision_test.go#L299-L316

Originally, there were written as part of dynamic provisioning. I cannot make out a reason as to why this behaviour was chosen.
https://github.com/kubernetes/kubernetes/pull/29006

Default storage classes were added later, and I cannot find again the reason for this behavior:
https://github.com/kubernetes/kubernetes/pull/30900

I would expect that storageClass is not overwritten by dynamic provisioner only if the associated PV was created manually.
I don't see a use-case for using an empty value, _if_ the claim is not bound and doesn't specify a persistentVolume.
Even for manual use-cases, the usual value that's typically used for the storageClass would be "none" or "local".
Does this make sense to you or am I missing something?

**Comment by msau42** [comment]
> Is this really something we want to do? If I recall correctly, having no class is semantically meaningful, isn't it?

I think we definitely don't want to set the default storageclass for PVCs that have already been bound.

For unbound PVCs, I see two scenarios:
1. Someone deployed all their components in parallel, including default StorageClass and workloads that use it. The ordering constraint currently imposed means that sometimes the default storageclass doesn't get set on those PVCs and they're stuck forever in an unbound state. This forces folks to order their deployments or hardcode a storageclass name in their PVC. One could argue that we have this ordering issue with other objects too, like CRDs and CRs.
2. Someone does not have a default storageclass and they manually provide PVs for the PVCs. In this case, a nil PVC.storageclassname will bind to the PV. This is the scenario that may be impacted if we make this change. Note that it only breaks if a default storageclass gets installed in the cluster at some later point. The sequence that may break is:
    1. User creates PVC with empty StorageClass
    2. User creates a default StorageClass
    3. User creates a PV with empty StorageClass

IMO the scenario of depending on this behavior before installing a default StorageClass in the system seems a bit contrived and requires careful deployment ordering by the user. It could be mitigated by creating the PVC with "" storageclass.

That being said, I do agree that this makes explaining the API semantics more complicated. There are other issues too, such as [validation not allowing updates to storageclassname](https://github.com/kubernetes/kubernetes/blob/fb3273774ad0738fadd5a18693741d6818187b65/pkg/apis/core/validation/validation.go#L2088), and maybe some concerns (that I haven't thought about yet) with pvc templates (statefulset, generic ephemeral volumes).

In summary, this is not as simple to change, we need to think more deeply on the api compatibility and impact to other features.

**Comment by LittleChimera** [comment]
> 2. Someone does not have a default storageclass and they manually provide PVs for the PVCs. In this case, a nil PVC.storageclassname will bind to the PV. This is the scenario that may be impacted if we make this change. Note that it only breaks if a default storageclass gets installed in the cluster at some later point. The sequence that may break is:
>
>    1. User creates PVC with empty StorageClass
>    2. User creates a default StorageClass
>    3. User creates a PV with empty StorageClass

It's a good point. But, I think we can break this down further in two scenarios:

   1. User expects a default storage class present on their system. In this case, if you want to provide PV manually, you're expected to define some storage class - e.g. "", "none", "manual". Otherwise, the storageClass will be set to default one by the mutating admission webhook.
   2. User doesn't expect a default storage class present on their system. In this case, the change would be irrelevant. Adding a default StorageClass at a much later stage would not make a difference because a PV will be bound to the PVC.

I think it should be enough to change the logic so that storageClass can change only for PVCs with nil value. I.e. users that don't specify storage class on their PVC, expect that a default one will be set eventually. Users who want to provision PVs manually fall into one of the two scenarios above which both seem safe to me. Additionally, this might require changes to defaulter I think, so that might cause additional issues.

Do you think that makes sense and is worth a try?

> That being said, I do agree that this makes explaining the API semantics more complicated. There are other issues too, such as [validation not allowing updates to storageclassname](https://github.com/kubernetes/kubernetes/blob/fb3273774ad0738fadd5a18693741d6818187b65/pkg/apis/core/validation/validation.go#L2088), and maybe some concerns (that I haven't thought about yet) with pvc templates (statefulset, generic ephemeral volumes).

I didn't know about generic ephemeral volumes before, but from what I've read I don't see any clear issues for them or STS provisoned PVCs.
I still agree there might be some other concerns that we're missing.

**Comment by jsafrane** [comment]
> Is this really something we want to do?

I think that issues described my @msau42 are real, we have the same ordering problem too. Similarly, it's not possible to reliably change the default storage class - most fields are read-only and currently we must delete the storage class and re-create it later, hoping that no PVC is created in the meantime.

>  If I recall correctly, having no class is semantically meaningful, isn't it?

There is a difference between `storageClassName: ""` (*PVC does not want any storage class*) and `storageClassName: nil` (*PVC wants the default one and it falls back to `""` if there is no default storage class at the point when PVC was created*). IMO, it makes sense to extend `nil` to "*PVC wants the default storage class, even if it's created later*".

If I remember correctly, with fall back from `nil` to `""` we wanted to keep the existing PVC behavior from time when there were no storage classes at all.

I also remember that we used annotation + admission plugin for the default storage class to have possibility to have different defaulting rules / admission plugins in the future, e.g. a different default class for different namespaces. If this was real, then we cannot implement the defaulting in PV controller, because it does not know what the defaulting rules are. But I haven't seen these different defaulting rules used in the wild, IMO everyone uses our admission plugin.

**Comment by dims** [comment]
please reopen if you still need this WIP PR. Trying to declutter the open PR(s)

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 3]
```diff
@@ -359,8 +360,19 @@ func (ctrl *PersistentVolumeController) syncUnboundClaim(claim *v1.PersistentVol
 					return err
 				}
 				return nil
+			case storagehelpers.GetPersistentVolumeClaimClass(claim) == "":
```
[the remark this answers is not in the document]
Comment (mkimuram) [reply pkg/controller/volume/persistentvolume/pv_controller.go:363]: I've added some codes on top of this PR in order to:
- Work (fix wrong error check, loosen validation)
- Avoid trying to add default StorageClass when "" is intentionally set by users (although, it is actually blocked by validation)
- Only allow one default StorageClass

https://github.com/kubernetes/kubernetes/compare/master...mkimuram:storage-class-default-pv-controller-v3

It would be better tighten validation a bit more, but there won't be a big user facing change.
